Mathlib/Topology/Germ.lean

```diff
@@ -0,0 +1,190 @@
+      intro f f' hff' hf
+      apply (hf.and <| Eventually.eventually_nhds hff').mono
+      rintro y ⟨hy, hy'⟩
+      rwa [Germ.coe_eq.mpr (EventuallyEq.symm hy')]
+    fun f f' hff' ↦ propext <| forall_congr' fun _ ↦ ⟨this f f' hff', this f' f hff'.symm⟩
+
+theorem Filter.Eventually.germ_congr
+    {P : Germ (𝓝 x) Y → Prop} (hf : P f) (h : ∀ᶠ z in 𝓝 x, g z = f z) : P g := by
+  convert hf using 1
+  exact Germ.coe_eq.mpr h
+
+theorem Filter.Eventually.germ_congr_set
+    {P : ∀ x : X, Germ (𝓝 x) Y → Prop} (hf : ∀ᶠ x in 𝓝ˢ A, P x f)
+    (h : ∀ᶠ z in 𝓝ˢ A, g z = f z) : ∀ᶠ x in 𝓝ˢ A, P x g := by
+  rw [eventually_nhdsSet_iff_forall] at *
+  intro x hx
+  apply ((hf x hx).and (h x hx).eventually_nhds).mono
+  exact fun y hy ↦ hy.2.germ_congr hy.1
+
+theorem restrictGermPredicate_congr {P : ∀ x : X, Germ (𝓝 x) Y → Prop}
+    (hf : RestrictGermPredicate P A x f) (h : ∀ᶠ z in 𝓝ˢ A, g z = f z) :
+    RestrictGermPredicate P A x g := by
+  intro hx
+  apply ((hf hx).and <| (eventually_nhdsSet_iff_forall.mp h x hx).eventually_nhds).mono
+  rintro y ⟨hy, h'y⟩
+  rwa [Germ.coe_eq.mpr h'y]
+
+theorem forall_restrictGermPredicate_iff {P : ∀ x : X, Germ (𝓝 x) Y → Prop} :
+    (∀ x, RestrictGermPredicate P A x f) ↔ ∀ᶠ x in 𝓝ˢ A, P x f := by
+  rw [eventually_nhdsSet_iff_forall]
+  rfl
+
+theorem forall_restrictGermPredicate_of_forall
+    {P : ∀ x : X, Germ (𝓝 x) Y → Prop} (h : ∀ x, P x f) :
+    ∀ x, RestrictGermPredicate P A x f :=
+  forall_restrictGermPredicate_iff.mpr (eventually_of_forall h)
+end RestrictGermPredicate
+
+namespace Filter.Germ
+/-- Map the germ at of functions `X × Y → Z` at `p=(x,y) ∈ X × Y` to the corresponding germ
+  of functions `X → Z` at `x ∈ X` -/
+def sliceLeft [TopologicalSpace Y] {p : X × Y} (P : Germ (𝓝 p) Z) : Germ (𝓝 p.1) Z :=
+  P.compTendsto (Prod.mk · p.2) (Continuous.Prod.mk_left p.2).continuousAt
+
+@[simp]
+theorem sliceLeft_coe [TopologicalSpace Y] {y : Y} (f : X × Y → Z) :
+    (↑f : Germ (𝓝 (x, y)) Z).sliceLeft = fun x' ↦ f (x', y) :=
+  rfl
+
+/-- Map the germ at of functions `X × Y → Z` at `p=(x,y) ∈ X × Y` to the corresponding germ
+  of functions `Y → Z` at `y ∈ Y` -/
+def sliceRight [TopologicalSpace Y] {p : X × Y} (P : Germ (𝓝 p) Z) : Germ (𝓝 p.2) Z :=
+  P.compTendsto (Prod.mk p.1) (Continuous.Prod.mk p.1).continuousAt
+
+@[simp]
+theorem sliceRight_coe [TopologicalSpace Y] {y : Y} (f : X × Y → Z) :
+    (↑f : Germ (𝓝 (x, y)) Z).sliceRight = fun y' ↦ f (x, y') :=
+  rfl
+
+/-- The germ of functions `X → Y` at `x ∈ X` is constant w.r.t. the neighbourhood filter `𝓝 x`. -/
+def IsConstant (P : Germ (𝓝 x) Y) : Prop :=
+  P.liftOn (fun f ↦ ∀ᶠ x' in 𝓝 x, f x' = f x) <| by
+    suffices : ∀ f g : X → Y, f =ᶠ[𝓝 x] g → (∀ᶠ x' in 𝓝 x, f x' = f x) → ∀ᶠ x' in 𝓝 x, g x' = g x
+    exact fun f g hfg ↦ propext ⟨fun h ↦ this f g hfg h, fun h ↦ this g f hfg.symm h⟩
+    rintro f g hfg hf
+    refine (hf.and hfg).mono fun x' hx' ↦ ?_
+    rw [← hx'.2, hx'.1, hfg.eq_of_nhds]
```

Comment:
This could be defined for any filter, right? Of course you wouldn't have the distinguished value `x` so the proof might get slightly more complicated, but I think this is something that we want in general.

Reply:
Good point. In fact, my new definition makes the proofs slightly *simpler*. :-)